src/java.base/share/classes/jdk/internal/crac/JDKContext.java

```diff
@@ -26,25 +26,96 @@
 
 package jdk.internal.crac;
 
+import jdk.crac.CheckpointException;
+import jdk.crac.Context;
+import jdk.crac.Resource;
+import jdk.crac.RestoreException;
 import jdk.crac.impl.AbstractContextImpl;
+import jdk.internal.access.JavaIOFileDescriptorAccess;
+import jdk.internal.access.SharedSecrets;
+import sun.security.action.GetBooleanAction;
 
-import java.util.Comparator;
-import java.util.Map;
+import java.io.File;
+import java.io.FileDescriptor;
+import java.io.IOException;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.util.*;
+import java.util.stream.Collectors;
 
 public class JDKContext extends AbstractContextImpl<JDKResource, Void> {
-    static class ContextComparator implements Comparator<Map.Entry<JDKResource, Void>> {
-        @Override
-        public int compare(Map.Entry<JDKResource, Void> o1, Map.Entry<JDKResource, Void> o2) {
-            return o1.getKey().getPriority().compareTo(o2.getKey().getPriority());
+    public static final String COLLECT_FD_STACKTRACES_PROPERTY = "jdk.crac.collect-fd-stacktraces";
+    public static final String COLLECT_FD_STACKTRACES_HINT = "Use -D" + COLLECT_FD_STACKTRACES_PROPERTY + "=true to find the source.";
+
+    // JDKContext by itself is initialized too early when system properties are not set yet
+    public static class Properties {
+        public static final boolean COLLECT_FD_STACKTRACES =
+                GetBooleanAction.privilegedGetProperty(JDKContext.COLLECT_FD_STACKTRACES_PROPERTY);
+    }
+
+    private WeakHashMap<FileDescriptor, Object> claimedFds;
+
+    public boolean matchClasspath(String path) {
+        Path p = Path.of(path);
+        String classpath = System.getProperty("java.class.path");
+        int index = 0;
+        while (index >= 0) {
```

Comment:
Bug: looks like the update of `index` was dropped somewhere, this loop may run infinitely...